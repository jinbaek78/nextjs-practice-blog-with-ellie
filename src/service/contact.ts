import { EmailData } from './email';

export async function sendContactEmail(email: EmailData) {
  const response = await fetch('http://localhost:3000/api/contact', {
    method: 'POST',
    body: JSON.stringify(email),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  const data = await response.json();
  if (!response.ok) {
    throw new Error(data.message || 'failed to request to server');
  }

  return data;
}
