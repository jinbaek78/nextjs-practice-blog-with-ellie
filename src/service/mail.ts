import { BannerData } from '@/components/Banner';
import { Form } from '@/components/ContactForm';
import { object, string, number, date, InferType } from 'yup';

export async function sendEmail(
  data: Form
  // callback: (banner: BannerData) => void
): Promise<BannerData> {
  let formSchema = object({
    subject: string().required(),
    from: string().email().required(),
    message: string().min(2).required(),
  });

  let dataValidationResult;
  try {
    dataValidationResult = await formSchema.validate(data);
  } catch (err) {
    console.log(err);
    return { message: 'please, write the right format', state: 'error' };
  }

  const formHeader = new Headers();
  formHeader.append('Content-Type', 'application/json');

  const requestOptions = {
    method: 'POST',
    headers: formHeader,
    body: JSON.stringify(data),
  };

  let res;
  let resultBanner: BannerData;
  try {
    res = await fetch('http://localhost:3000/api/contact', requestOptions);
  } catch (err) {
    console.log(err);
    return (resultBanner = { message: 'fail', state: 'error' });
  }

  return res.status === 200
    ? { message: 'success', state: 'success' }
    : { message: 'fail', state: 'error' };

  // return fetch('http://localhost:3000/api/contact', requestOptions)
  //   .then((response) => response.json())
  //   .then((result) => console.log(result))
  //   .catch((error) => console.log('error', error));
}
