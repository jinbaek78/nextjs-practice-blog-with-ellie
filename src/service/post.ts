import fs from 'fs/promises';
import path from 'path';
import { cwd } from 'process';

export type Post = {
  title: string;
  description: string;
  date: string;
  category: string;
  path: string;
  featured: boolean;
};

export async function getPostsData(): Promise<Post[]> {
  const postsData = await JSON.parse(
    await fs.readFile(`${path.join(cwd(), 'public', 'data', 'posts.json')}`, {
      encoding: 'utf-8',
    })
  );

  return postsData;
}
