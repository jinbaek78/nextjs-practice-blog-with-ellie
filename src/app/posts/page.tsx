import Posts from '@/components/Posts';
import { getAllPosts } from '@/service/posts';

export default async function PostsPage() {
  const posts = await getAllPosts();

  return <Posts posts={posts} />;
}
