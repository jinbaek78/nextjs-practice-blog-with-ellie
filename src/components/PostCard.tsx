import { Post } from '@/service/post';
import Image from 'next/image';

type Props = {
  postData: Post;
};
export default function PostCard({
  postData: { path, date, title, description, category },
}: Props) {
  console.log(path, date, title);
  return (
    <li className="flex flex-col gap-1 shadow-xl rounded-lg">
      <Image
        className="rounded-lg"
        src={`/images/posts/${path}.png`}
        alt="post-image"
        width={400}
        height={150}
      />
      <p className="text-right text-zinc-800 mr-3">{date}</p>
      <div className="flex flex-col justify-center items-center">
        <h3 className="font-semibold text-lg">{title}</h3>
        <h5 className="text-sm">{description}</h5>
        <div className="bg-green-200 rounded-xl py-1 px-4 text-sm mt-3 mb-5">
          {category}
        </div>
      </div>
    </li>
  );
}
