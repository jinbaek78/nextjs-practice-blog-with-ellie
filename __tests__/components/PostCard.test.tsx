import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import PostData from '../../public/data/posts.json';
import { Post } from '@/service/posts';
import PostCard from '@/components/PostCard';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import mockRouter from 'next-router-mock';

describe.skip('PostCard', () => {
  const post: Post = PostData[0];
  const { title, description, date, category, path } = post;

  it('renders correctly', () => {
    const { container } = render(<PostCard post={post} />);
    expect(container).toMatchSnapshot();
  });

  it('renders post details correctly', () => {
    render(<PostCard post={post} />);

    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(description)).toBeInTheDocument();
    expect(screen.getByText(date.toString())).toBeInTheDocument();
    expect(screen.getByText(category)).toBeInTheDocument();
  });

  it('navigates to clicked post page on the post click', async () => {
    render(<PostCard post={post} />, { wrapper: MemoryRouterProvider });

    await fireEvent.click(screen.getByRole('link'));
    expect(mockRouter.asPath).toBe(`/posts/${path}`);
  });
});
