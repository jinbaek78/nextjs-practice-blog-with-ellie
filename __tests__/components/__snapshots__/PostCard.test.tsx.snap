// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard renders correctly 1`] = `
<div>
  <a
    href="/posts/review-2021"
  >
    <article
      class="rounded-md overflow-hidden shadow-md hover:shadow-xl"
    >
      <img
        alt="2021년 리뷰"
        class="w-full"
        data-nimg="1"
        decoding="async"
        height="200"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=640&q=75"
        srcset="/_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=384&q=75 1x, /_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=640&q=75 2x"
        style="color: transparent;"
        width="300"
      />
      <div
        class="flex flex-col items-center p-4"
      >
        <time
          class="self-end text-gray-700"
        >
          2021-12-31
        </time>
        <h3
          class="text-lg font-bold"
        >
          2021년 리뷰
        </h3>
        <p
          class="w-full truncate text-center"
        >
          2021년 한해 동안 배운 기술들, 생각들 정리
        </p>
        <span
          class="text-sm rounded-lg bg-green-100 px-2 my-2"
        >
          my story
        </span>
      </div>
    </article>
  </a>
</div>
`;
